src/main/java/io/github/moulberry/notenoughupdates/miscgui/itemcustomization/ItemCustomizationUtils.java

```diff
@@ -0,0 +1,344 @@
+		Gui.drawRect(xCenter - 90, yTop, xCenter + 92, yTop + 17, 0x70000000);
+		Gui.drawRect(xCenter - 90, yTop, xCenter + 90, yTop + 15, 0xff101016);
+		Gui.drawRect(xCenter - 89, yTop + 1, xCenter + 89, yTop + 14, 0xff000000 | colour);
+
+		Utils.renderShadowedString(text, xCenter, yTop + 4, 180);
+
+		if (renderReset) {
+			Minecraft.getMinecraft().getTextureManager().bindTexture(RESET);
+			GlStateManager.color(1, 1, 1, 1);
+			RenderUtils.drawTexturedRect(xCenter + 90 - 12, yTop + 2, 10, 11, GL11.GL_NEAREST);
+		}
+	}
+
+	public static void renderTextBox(
+		GuiElementTextField textField, String text, int xOffset, int yOffset, int maxTextSize
+	) {
+		if (!textField.getFocus() && textField.getText().isEmpty()) {
+			textField.setOptions(GuiElementTextField.SCISSOR_TEXT);
+			textField.setPrependText(text);
+		} else {
+			textField.setOptions(GuiElementTextField.COLOUR | GuiElementTextField.SCISSOR_TEXT);
+			textField.setPrependText("");
+		}
+
+		if (!textField.getFocus()) {
+			textField.setSize(maxTextSize, 20);
+		} else {
+			int textSize = Minecraft.getMinecraft().fontRendererObj.getStringWidth(textField.getTextDisplay()) + 10;
+			textField.setSize(Math.max(textSize, maxTextSize), 20);
+		}
+
+		textField.render(xOffset, yOffset);
+	}
+
+	public static Color getColourFromHex(String hex) {
+		Color color = null;
+		try {
+			int decode = Integer.decode(hex);
+			color = new Color(decode);
+		} catch (NumberFormatException | NullPointerException e) {
+		}
+		return color;
+	}
+
+	public static int rgbToInt(Color color) {
+		return (color.getRed() << 16) | (color.getGreen() << 8) | color.getBlue();
+	}
+
+	public static void renderPresetButtons(int x, int y, boolean valid, boolean secondValid, String preset) {
+		Minecraft.getMinecraft().getTextureManager().bindTexture(custom_ench_colour);
+		GlStateManager.color(1, 1, 1, 1);
+		Utils.drawTexturedRect(
+			x - 88 + 198,
+			y + 2,
+			88,
+			20,
+			64 / 217f,
+			152 / 217f,
+			48 / 78f,
+			68 / 78f,
+			GL11.GL_NEAREST
+		);
+		Utils.drawTexturedRect(
+			x - 88 + 198,
+			y + 2 + 24,
+			88,
+			20,
+			64 / 217f,
+			152 / 217f,
+			48 / 78f,
+			68 / 78f,
+			GL11.GL_NEAREST
+		);
+
+		Utils.drawStringCenteredScaledMaxWidth("Load " + preset, x - 44 + 198, y + 8, false, 86, 4210752);
+		Utils.drawStringCenteredScaledMaxWidth("from Clipboard", x - 44 + 198, y + 16, false, 86, 4210752);
+		Utils.drawStringCenteredScaledMaxWidth("Save " + preset, x - 44 + 198, y + 8 + 24, false, 86, 4210752);
+		Utils.drawStringCenteredScaledMaxWidth("to Clipboard", x - 44 + 198, y + 16 + 24, false, 86, 4210752);
+
+		if (!valid) {
+			Gui.drawRect(x - 88 + 198, y + 2, x + 198, y + 2 + 20, 0x80000000);
+		}
+		if (!secondValid) {
+			Gui.drawRect(x - 88 + 198, y + 2 + 24, x + 198, y + 2 + 20 + 24, 0x80000000);
+		}
+
+		GlStateManager.color(1, 1, 1, 1);
+	}
+
+	public static boolean validShareContents(String sharePrefix) {
```

Comment:
This function seems to quite often lock up the game for ~10 seconds or so. This might be an issue with this function being called multiple times in a single frame? Not sure, but it does always lock up on the second call to this method. Maybe we should invest in a mod wide clipboard cache (using .getTransferable, gets updated once per tick) (This is with my clipboard containing about 100 characters or so, and it locks up while waiting on obtaining a lock, so this is not about big data transfers)

Reply:
This has no performance impact for me, I assume it's an issue with the linux implementation. I think your clipboard idea should probably be in a separate pr

Reply:
Made a pr that should hopefully fix this